Refer to feature policy spec

clelland opened the pull request:
This change defines Fullscreen as a policy-controlled feature,
referencing the feature policy spec. It makes use of the currently-
monkey-patched "Allowed to use" algorithm in HTML to allow the
document's policy to control whether an element is permitted to go
fullscreen.

Fixes: #106


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/fullscreen/107.html" title="Last updated on Jul 6, 2018, 6:54 PM GMT (19ba697)">Preview</a> | <a href="https://whatpr.org/fullscreen/107/6e49fbc...19ba697.html" title="Last updated on Jul 6, 2018, 6:54 PM GMT (19ba697)">Diff</a>

foolip commented:
@clelland, I accidentally lost this in my inbox way back, sorry! I've rebased it and tried some tweaks locally. Sent https://github.com/WICG/feature-policy/pull/179 to export two definitions needed, can you merge that?

clelland commented:
Merged! (I thought those were exported already; thanks for fixing that)

clelland commented on fullscreen.bs:
```diff
@@ -629,6 +629,23 @@ iframe:fullscreen {
+
+<p>A <a>document</a>'s <a>feature policy</a> determines whether any content in that document is allowed to
+go fullscreen. If disabled in any document, no content in the document will be <a>allowed to use</a>
+fullscreen. If enabled, content in the document will be <a>allowed to use</a> fullscreen.
```
Yes, I have a PR at https://github.com/whatwg/html/pull/3287 which attempts to do just that. Working on a couple of final integration points before that can (hopefully) be merged.

foolip replied:
Should this PR be put on pause for a bit while that is finalized?

clelland replied:
Now that that has landed, I've updated the PR to be more in line with it.

clelland commented:
@foolip, can you take a look? I've updated the wording in the PR to be more like what was used in XHR, and to work with the changes to HTML. I also changed what is really a non-normative description of the way that `allow` and `allowfullscreen` work into a `<div class="note">`.

clelland commented:
Yes, fullscreen is pretty well tested, being one of the first features that Chrome shipped with -- these tests all check the behaviour of `fullscreen *` in policies:

* [feature-policy-frame-policy-allowed-for-all.https.sub.html](https://wpt.fyi/results/feature-policy/feature-policy-frame-policy-allowed-for-all.https.sub.html)
* [feature-policy-header-policy-allowed-for-all.https.sub.html](https://wpt.fyi/results/feature-policy/feature-policy-header-policy-allowed-for-all.https.sub.html)
* [feature-policy-nested-header-policy-allowed-for-all.https.sub.html](https://wpt.fyi/results/feature-policy/feature-policy-nested-header-policy-allowed-for-all.https.sub.html)
* [feature-policy-nested-header-policy-allowed-for-self.https.sub.html](https://wpt.fyi/results/feature-policy/feature-policy-nested-header-policy-allowed-for-self.https.sub.html)

They do all rely on the [`document.policy`](https://drafts.featurepolicy.rocks/add-introspection/#introspection) object which isn't shipped yet, though, so it may be worthwhile to add a couple of extra tests for that specifically.

(In general, I'd like to get to a point where we can test the feature policy mechanism itself separately from its integration with each feature -- with the goal being that if you can test that the feature works correctly when enabled and when disabled, then the wpt/feature-policy tests will ensure that all of the different possible combinations of frames and policies work to get the right {enabled,disabled} state down to the frame. For now, though, I don't mind adding tests for specific combinations of policy-structure+feature to increase test coverage.)

foolip commented:
Ah, so those tests all fail. But they do cover the bits changed in this PR, so I'll go ahead and merge now. Thanks!

clelland commented:
Yes -- they pass in Chrome with --enable-experimental-web-platform-features, but not without.
Thanks!

foolip commented:
Ah, yes, it works!
https://wpt.fyi/results/feature-policy/feature-policy-frame-policy-allowed-for-all.https.sub.html?label=experimental
